### DISCUSSION TITLE
Provide Example App for NestJS with Apollo GraphQL and MikroORM (w/ clean Identity Map per Request)

### DISCUSSION THREAD

**Discussion by woltob** [mikro-orm/mikro-orm#2212] [General]
Hi B4nan, 

you're doing amazing work with Mikro ORM and I can see this rising to become the standard ORM for NodeJS w/ Typescript, especially with NestJS.

Since GraphQL APIs are becoming more and more common for new projects where front-ends would be tightly coupled with their respective back-end APIs, I was wondering whether you (with the help of some NestJS and Apollo experts), could provide an example app with a go to configuration and setup, especially with regards to a clean per-request Identity Map via the Async Storage of the newer Node frameworks.

Required features:
- NestJS, Mikro-ORM, Apollo GraphQL
- Mikro-ORM configuration with GraphQL (clean identity map per request, injection of mikro orm helpers/classes in NestJS).
- GraphQL Error Handling where exceptions would be populated accordingly.
- Token-based authentication and respective strategies with GraphQL (via Header)
- Combined entity models for GraphQL and MikroORM (we managed to do that with @annotators) 

I think this would be a good setup and base for many more project to start from.

Keep up the amazing work!

**Comment by B4nan** [comment]
Can't help much with GQL, never really used it myself, but I can see how this combinations is popular - would be indeed great to have such example app, more complete one similar to the realworld example app. 

Regarding `AsyncLocalStorage`, it is already used by default in v5 with the `RequestContext` helper. I don't plan to add more guides for v4, if something, will target v5.

Would be probably best to adjust some existing example project and just swap the ORMs (and adjust the ORM specific parts). Do you know of something usable? Otherwise I don't have a clue, most of the points you mentioned are something I never used myself, so can't provide best practices 🤷.

Will pin this for some time, maybe someone can help.

**Comment by briandiephuis** [comment]
After a long love-hate relation with TypeORM I'm now checking out MikroORM. I usually work with NestJS, GraphQL and TypeGraphQL (code-first approach), so I'll give that a try. I'm not aware of any specific example using these libraries, but I'll give it a go to create RealWorld like app (is there a GQL spec for this?) based on my experience. Together we should be able to figure this out :+1:  I'm personally interested in adding dataloaders scoped to a single request, so I'll try that too. 

I'll post here when I have something to show.

**Comment by woltob** [comment]
@briandiephuis are you using Postgres as the database?

**Comment by briandiephuis** [comment]
@woltob Yes, I refactored the example to Postgres :muscle: 

I also decided to put the authentication (sign in, sign out, refresh token and forced sign-out) outside of GraphQL. This has worked well for me in the past and seems to be best practice for token based auth. This allows us to use a http-only refresh cookie on the `/auth` route. Token based auth through GQL would make it so that the refresh cookies would _have_ to be sent on every request which doubles the size of auth-related stuff on each request. I'm interested in what you all think about this :) (More details in the readme)

It's a bit more work than expected, and far from finished, but I'm getting somewhere now. See: https://github.com/briandiephuis/nestjs-realworld-example-app

I found an example of using Mikro-ORM with TypeGraphQL here: https://github.com/MichalLytek/type-graphql/tree/master/examples/mikro-orm but it is built on a bare ApolloServer. There they fork/clone the identity manager for each request [here](https://github.com/MichalLytek/type-graphql/blob/e0b1aa0cf4b4d5e3dfcff95f52c30f9f51bea6ea/examples/mikro-orm/index.ts#L52). I found this in the docs:

```TypeScript
app.use((req, res, next) => {
  RequestContext.create(orm.em, next);
});
```

@B4nan How can we get access to the `orm` object to clone the EntityManager when using NestJS? I think I should add it to each gql request [here](https://github.com/briandiephuis/nestjs-realworld-example-app/blob/f591935459ae090c0b100c8b6505145aef1a1b60/src/app.module.ts#L23)

**Comment by briandiephuis** [comment]
To elaborate on the problem; currently lazy loaded relations do not work:

The `user.entity` has this relation;

```ts
  @Field(() => [Article])
  @OneToMany(() => Article, (article) => article.author)
  articles = new Collection<Article>(this);
```

Pretty much exactly as shown here: https://github.com/MichalLytek/type-graphql/blob/master/examples/mikro-orm/entities/recipe.ts#L22

I'm planning of implementing dataloaders, and using lazy-loaded relations is usually a bad practise - but they are super useful for fast prototyping. Plus, dataloaders also require a request scoped entity manager.

<details>
  <summary>To reproduce</summary>
  
Create a User:
```gql
mutation createUser($input: CreateUserInput!) {
  createUser(input: $input) {
    id
  }
}
```
With variables:
```json
{
	"input": {
		"email": "d.duck@example.com",
		"password": "daisy01",
		"username": "Donald"
	}
}
```
Then sign in (POST to `http://localhost:8000/auth`) with JSON body:
```JSON
{
	"email": "d.duck@example.com",
	"password": "daisy01"
}
```

Create an article using that auth token

```gql
mutation createArticle($input: CreateArticleInput!) {
  createArticle(input: $input) {
    id
    title
    body
    description
  }
}
```
```JSON
{
	"input": {
		"body": "Article body",
		"description": "Awesome article",
		"title": "Awesome article",
		"tagList": [
			"article",
			"awesome"
		]
	}
}
```

Then request those, the user does not show any articles;
```gql
query me {
  me {
    id
    username
    image
    bio
    articles {
      id
      title
      description
    }
  }
}
```

Response:

```json
{
  "data": {
    "me": {
      "id": "1",
      "username": "Donald",
      "image": "",
      "bio": "",
      "articles": [] // <- should show the created article here (it is in the database with the correct link to the user)
    }
  }
}
```

</details>

**Comment by B4nan** [comment]
Be sure to read this: https://github.com/mikro-orm/nestjs/issues/22

Also as I already said, I really can't help much with GQL, so if you want help, please try to narrow it down to something that is not GQL related. 

Also worth reading:
https://jenyus.web.app/blog/2021-03-07-nestjs-starter

**Comment by notmedia** [comment]
Also having troubles with getting orm object for get clear identity map on each request.

Like the docs says, we should use middleware:

```
app.use((req, res, next) => {
  RequestContext.create(orm.em, next);
});
```

It's not a GQL problem, with Nest we have injected MikroORM Module and I don't understand how I can get `orm` object outside to use it in middleware. @B4nan can you explain please?

**Comment by notmedia** [comment]
Hm.. maybe I found the solution

@briandiephuis try this:

```ts
import { MikroORM } from '@mikro-orm/core';

import { AppModule } from './app.module';
import { storage } from './storage';

async function bootstrap() {
  const app = await NestFactory.create(AppModule);
  const orm = app.get(MikroORM);

  app.use((req, res, next) => {
    storage.run(orm.em.fork(true, true), next);
  });
 
 // or without AsyncLocalStorage
 
 app.use((req, res, next) => {
  RequestContext.create(orm.em, next);
 });
}
```

**Comment by briandiephuis** [comment]
Thanks for the tip @notmedia . That seems to work, and I've also found a more NestJS way by formatting your solution as NestJS middleware.

```ts
@Injectable()
export class MikroOrmRequestContextMiddleware implements NestMiddleware {
  constructor(private readonly orm: MikroORM) {}

  use(req: AppRequest, res: Response, next: NextFunction): void {
    RequestContext.create(this.orm.em, next);
  }
}
```

Now both dataloaded relations and more naively loaded relations work. Without pagination, though. That'll be a whole other story.

Getting auto-loaded relations to work seems to be impossible. But like I said, it's basically only useful for quick prototyping so I'm not too sad about that.

I'm swamped with work but I'll try to finish this example app :smile:

**Comment by notmedia** [comment]
Yeah, this approach is better

**Comment by rathax** [comment]
has anyone got a per-request Identity Map working when using nestJs + graphql with the fastify adapter?

**Comment by notmedia** [comment]
@rathax have you tried our approaches we posted above? If there is a bug, could you please post it here?

**Comment by rathax** [comment]
@notmedia 

yes, that's why I have asked if anyone got it working with fastify or if I just have missed something.

**Comment by notmedia** [comment]
@rathax I can't help you without any information

**Comment by jamesmeneghello** [comment]
We've got a commercial project that has NestJS set up with MikroORM / GraphQL with extensive support for authn/z, caching, cursor/offset pagination, extremely flexible query conditions, etc... I can't share the code, but I can talk about bits of it if anyone has questions.

e.g.: we don't use a dataloader, instead we break down the GraphQL query to determine which relations need loading and populate that all-at-once, instead - basically a heavily modified version of this: https://github.com/driescroons/graphql-fields-to-relations

**Comment by B4nan** [comment]
> e.g.: we don't use a dataloader, instead we break down the GraphQL query to determine which relations need loading and populate that all-at-once, instead - basically a heavily modified version of this: driescroons/graphql-fields-to-relations

What about contributing back to that library, or releasing your own version? :] Or an article about how you do things. I am quite sure all of that would be quite welcome.

**Comment by woltob** [comment]
Hi @jamesmeneghello , 

that sounds amazing. Perhaps you can publish a stripped down version of your code and setup (especially the configuration). 
I, and I guess many others, would be very interested in comparing and discussing solutions.
Also, a Medium article with code bits and pieces would be much appreciated.

Such a barebone boilerplate project would be an amazing starting point for others to venture out and build something amazing.

Let us know, thanks!

**Comment by jamesmeneghello** [comment]
Let me have a think about the best way to disseminate it - it'll probably take a couple of weeks to write up anyway, and it's pretty opinionated, but it has some fairly cool stuff in it.

**Comment by driescroons** [comment]
@jamesmeneghello Would love to have a talk about this! I'd love to elaborate on the repo, especially when we're getting closer to v5.

**Comment by ChiefJacob** [comment]
@jamesmeneghello  @briandiephuis Have you found a solution, or is there continued reading on this?

**Comment by b-bellovic** [comment]
I tried to go through all discussions and materials on github, nest.js, mikro-orm pages but I am still not able to make RequestContext work. If anybody can tell me what am I missing I would be really thankful. 

Here are files

_main.ts_
```ts
async function bootstrap() {
  const app = await NestFactory.create(AppModule, { bodyParser: false });
  const orm = app.get(MikroORM);

  app.use(bodyParser.json());
  app.use((req, res, next) => {
    RequestContext.create(orm.em, next);
  });
  app.enableShutdownHooks();
  await app.listen(3000);
}

bootstrap();
```

_app.module.ts_
```ts
@Module({
  imports: [
    ConfigModule.forRoot({ isGlobal: true }),
    MikroOrmModule.forRoot(),
    GraphQLModule.forRootAsync<ApolloDriverConfig>({
      driver: ApolloDriver,
      imports: [ConfigModule],
      useFactory: async (configService: ConfigService) => ({
        autoSchemaFile: join(process.cwd(), 'src/schema.gql'),
        playground: false,
        sortSchema: true,
        bodyParserConfig: false,
        context: ({ req, res }) => ({ req, res }),
        path: 'graphql',
        cors: {
          origin: getCorsOrigins(configService),
          credentials: true,
        },
        plugins: [ApolloServerPluginLandingPageLocalDefault()],
      }),
      inject: [ConfigService],
    }),
    UserModule,
  ],
})
export class AppModule {}
```

_mikro-orm.config.ts_
```ts
const logger = new Logger('MikroORM');

const config = {
  entities: ['dist/**/*.entity.js'],
  entitiesTs: ['src/**/*.entity.ts'],
  clientUrl: process.env.DATABASE_URL,
  type: 'postgresql',
  highlighter: new SqlHighlighter(),
  debug: true,
  logger: logger.log.bind(logger),
  metadataProvider: TsMorphMetadataProvider,
  // registerRequestContext: false, // doesn't make difference
  migrations: {
    snapshot: false,
    tableName: 'migrations', // name of database table with log of executed transactions
    path: './migrations', // path to the folder with migrations
    pattern: /^[\w-]+\d+\.ts$/, // regex pattern for the migration files
    transactional: true, // wrap each migration in a transaction
    disableForeignKeys: false, // wrap statements with `set foreign_key_checks = 0` or equivalent
    allOrNothing: true, // wrap all migrations in master transaction
    emit: 'ts', // migration generation mode
  },
  seeder: {
    path: './seeders', // path to the folder with migrations
    glob: '!(*.d).{js,ts}', // how to match seeder files (all .js and .ts files, but not .d.ts)
    emit: 'ts', // seeder generation mode
    fileName: (className: string) => className, // seeder file naming convention
  },
};

export default config;
```

_example query_
```ts
@Injectable()
export class UserQuery {
  constructor(private readonly em: EntityManager) {}

  @Query(() => User)
  user(@Args('id') id: string) {
    return this.em.findOne(User, { id });
  }
}

```

Result of query:
```
[Nest] 93024  - 07/25/2022, 7:21:03 PM   ERROR [ExceptionsHandler] Using global EntityManager instance methods for context specific actions is disallowed. If you need to work with the global instance's identity map, use `allowGlobalContext` configuration option or `fork()` instead.

```

**Comment by carmona44** [comment]
Hello guys! Any update on this? I'm interested in use NestJS + GraphQL + PostgreSQL + MikroORM. I think it could be a good approach for many projects, but I haven't found much information about this combination.

**Reply by B4nan** [reply]
Latest addition to the example repos is https://github.com/Cloudshelf/Shopify_CSConnector which uses this stack. I haven't look too deep into it, but its quite a fat/realworld example app.

**Reply by carmona44** [reply]
Looks good, I'll give it a try! Thank you very much for your quick response
